util/profile/flamegraph.go

```diff
@@ -0,0 +1,182 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package profile
+
+import (
+	"fmt"
+	"math"
+	"sort"
+
+	"github.com/google/pprof/profile"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/texttree"
+)
+
+type flamegraphNode struct {
+	cumValue int64
+	children map[uint64]*flamegraphNode
+}
+
+func newFlamegraphNode() *flamegraphNode {
+	return &flamegraphNode{
+		cumValue: 0,
+		children: make(map[uint64]*flamegraphNode),
+	}
+}
+
+// add the value from a sample into the flamegraph DAG.
+// This method should only be called on the root node.
+func (n *flamegraphNode) add(sample *profile.Sample) {
+	// FIXME: we take the last sample value by default, but some profiles have multiple samples.
+	//  - allocs:    alloc_objects, alloc_space, inuse_objects, inuse_space
+	//  - block:     contentions, delay
+	//  - cpu:       samples, cpu
+	//  - heap:      alloc_objects, alloc_space, inuse_objects, inuse_space
+	//  - mutex:     contentions, delay
+
+	value := sample.Value[len(sample.Value)-1]
+	if value == 0 {
+		return
+	}
+
+	locs := sample.Location
+
+	for {
+		n.cumValue += value
+		if len(locs) == 0 {
+			return
+		}
+
+		// The previous implementation in TiDB identify nodes using location ID,
+		// but `go tool pprof` identify nodes using function ID. Should we follow?
+		loc := locs[len(locs)-1].ID
+		child, ok := n.children[loc]
+		if !ok {
```

Comment:
I'm not sure whether is it proper for some sample contains recursive (or cumulate the value to existed children).

Reply:
Recursive call generates the same location at different depths (so we'll see ... → A → A → A → ... in the flamegraph).